### PR TITLE
Define and usse datatype for seat-seconds

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
The first commit of this PR defines a fixed-point datatype for seat-seconds.

The second commit of this PR revises the queueset code to use the new datatype in place of `float64`.  This includes a little relaxation of one queueset unit test to keep it passing, and making the staggering of events larger so that it does not get lost in the truncation.  This PR also renames the fields that held seat-seconds values.

The third commit adds code to prevent overflow of the global progress meter R.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
This is an alternative to #105322 , defining a datatype for the new representation.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
I experimented with different values of `ssScale`.
- `1e10` made all the existing tests pass.  TestUniformFlowsHandSize3 got average occupancies of   1.9999999998166644 and  1.9999999997499978.
- `1e9` made all the existing tests pass.  TestUniformFlowsHandSize3 got averages occupancies of  2.016666666449997 and 1.9833333331166643.
- `1e8` made all tests pass except TestUniformFlowsHandSize3.  It got average occupancies of  1.9499999999999988 and 2.049999999566666.  It had a margin of 0.01, meaning that the range of acceptable averages was 1.98--2.02.  I changed the margin to 0.03, making the range 1.94--2.06.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go [hunk 1]
```diff
@@ -134,3 +137,54 @@ func (q *queue) dump(includeDetails bool) debug.QueueDump {
...
+// no other code should use them.
+type SeatSeconds uint64
+
+// MaxSeatsSeconds is the maximum representable value of SeatSeconds
+const MaxSeatSeconds = SeatSeconds(math.MaxUint64)
+
+// SeatsTimesSeconds produces the SeatSeconds value for the given factors.
+// This is intended only to produce small values, increments in work
+// rather than amount of work done since process start.
+func SeatsTimesSeconds(seats, seconds float64) SeatSeconds {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:162]: The whole point I would like to use int64 (or uint64) instead of float64 is that I would like to get the int-related precision.

I would like to get rid of this function [and leave only the one below - duration is implemented using int...]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go [hunk 2]
```diff
@@ -134,3 +137,54 @@ func (q *queue) dump(includeDetails bool) debug.QueueDump {
...
+// SeatsTimeDuration produces the SeatSeconds value for the given factors.
+// This is intended only to produce small values, increments in work
+// rather than amount of work done since process start.
+func SeatsTimesDuration(seats float64, duration time.Duration) SeatSeconds {
+	return SeatSeconds(math.Round(seats * float64(duration/time.Nanosecond) / (1e9 / ssScale)))
+}
+
+// ToFloat converts to a floating-point representation.
+// This conversion may lose precision.
+func (ss SeatSeconds) ToFloat() float64 {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:175]: Same here - can we get rid of this function?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go [hunk 3]
```diff
@@ -134,3 +137,54 @@ func (q *queue) dump(includeDetails bool) debug.QueueDump {
...
+func (ss SeatSeconds) String() string {
+	const div = SeatSeconds(ssScale)
+	quo := ss / div
+	rem := ss - quo*div
+	quoStr := strconv.FormatUint(uint64(quo), 10)
+	if rem == 0 {
+		return quoStr + ".0ss"
+	}
+	remStr := strconv.FormatUint(uint64(rem), 10)
+	last := strings.LastIndexAny(remStr, "123456789")
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:190]: Actually - why do we need this?

I think it would be much easier to follow if we would be always printing exactly ssScaleDigits.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 4]
```diff
@@ -102,9 +102,9 @@ type queueSet struct {
 	// queues are still draining.
 	queues []*queue
 
-	// virtualTime is the amount of seat-seconds allocated per queue since process startup.
+	// currentR is the amount of seat-seconds allocated per queue since process startup.
 	// This is our generalization of the progress meter named R in the original fair queuing work.
-	virtualTime float64
+	currentR SeatSeconds
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:107]: It's not that I'm opposed to this change - in fact I'm 100% supportive.
But mixing different things within a single commit (i.e. renaming of variables, logic etc. makes it much harder to review it...)

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 5]
```diff
@@ -707,12 +707,12 @@ func (qs *queueSet) findDispatchQueueLocked() *queue {
 		queue := qs.queues[qs.robinIndex]
 		oldestWaiting, _ := queue.requests.Peek()
 		if oldestWaiting != nil {
-			sMin = math.Min(sMin, queue.virtualStart)
-			sMax = math.Max(sMax, queue.virtualStart)
-			estimatedWorkInProgress := qs.estimatedServiceSeconds * float64(queue.seatsInUse)
-			dsMin = math.Min(dsMin, queue.virtualStart-estimatedWorkInProgress)
-			dsMax = math.Max(dsMax, queue.virtualStart-estimatedWorkInProgress)
-			currentVirtualFinish := queue.virtualStart + qs.estimatedServiceSeconds*float64(oldestWaiting.Seats())
+			sMin = math.Min(sMin, queue.nextDispatchR.ToFloat())
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:752]: Please don't convert to floats - the whole point I want this to be int is to not loose prevision on math operations.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 6]
```diff
@@ -409,10 +409,52 @@ func (qs *queueSet) syncTimeLocked() {
 	realNow := qs.clock.Now()
 	timeSinceLast := realNow.Sub(qs.lastRealTime)
 	qs.lastRealTime = realNow
-	qs.currentR += SeatsTimesDuration(qs.getVirtualTimeRatioLocked(), timeSinceLast)
+	prevR := qs.currentR
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:412]: Can we move this commit to a separate PR?

I'm fine with that conceptually, but I think it requires a bit more attention.
And I would really like to unblock the other PRs (which are blocked on the other two commits) first.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:752]: `sMin` is here only to compute a Prometheus gauge value (which is necessarily a float64, right?).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 4 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:107]: There is no logic change here.

The only fields whose name change are fields whose type changed, so all the same lines of code get touched.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 6 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:412]: Yes.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:162]: By "get the int-related precision" do you mean the equivalent of setting `ssScale = 1e9` in this code?  Why is that particularly valuable?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:107]: I don't fully agree - most of the lines affected by rename are logs. Separating those would actually make this PR simpler.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 5 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:752]: I know - but still. I would like all this code to operate on int.

[And the only potential conversions to floats happening in the metrics code]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:162]: I just don't trust floats :)

even "add the same value and substract it may lead to getting different number".

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 4 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:107]: The logs have to be changed anyway.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go]: I eliminated SeatsTimesSeconds.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go (hunk 2 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:162]: No, because we need floating-point values to set Prometheus gauges.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go]: Deferred the conversion to floating-point to the metrics call.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go (hunk 3 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go]: Changed to always output all fraction digits.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 6 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:412]: See #105469

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:107]: Most of them are just changing only names - so I don't fully agree.
But I'm not going to push you for changing that - it's more a comment for the future.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/seatsecs_test.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+func TestSeatSecondsString(t *testing.T) {
+	digits := math.Log10(ssScale)
+	expectFmt := fmt.Sprintf("%%%d.%dfss", int(digits+2), int(digits))
+	testCases := []struct {
+		ss     SeatSeconds
+		expect string
+	}{
+		{ss: SeatSeconds(1), expect: fmt.Sprintf(expectFmt, 1.0/ssScale)},
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/seatsecs_test.go:33]: Let's not use fmt.Sprintf - let's use explicit constant string (as below).

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/seatsecs_test.go:33]: Let's address this in a subsequent PR - I'm going to let this PR go.
